File: client/src/game/units/Weapon.ts
import { convertColorNameToSideColor, SIDE_COLOR } from "@/utils/colors";

interface IWeapon {
  id: string;
  name: string;
  sideId: string;
  className: string;
  latitude: number;
  longitude: number;
  altitude: number;
  heading: number;
  speed: number;
  currentFuel: number;
  maxFuel: number;
  fuelRate: number; // lbs/hr
  range: number;
  route?: number[][];
  sideColor?: string | SIDE_COLOR;
  targetId: string | null;
  lethality: number;
  maxQuantity: number;
  currentQuantity: number;
}

export default class Weapon {
  id: string;
  name: string;
  sideId: string;
  className: string;
  latitude: number;
  longitude: number;
  altitude: number; // FT ASL -- currently default -- need to reference from database
  heading: number;
  speed: number; // KTS -- currently default -- need to reference from database
  currentFuel: number;
  maxFuel: number;
  fuelRate: number; // lbs/hr
  range: number; // NM -- currently default -- need to reference from database
  route: number[][];
  sideColor: SIDE_COLOR;
  targetId: string | null;
  lethality: number; // currently default -- need to reference from database
  maxQuantity: number;
  currentQuantity: number;

  constructor(parameters: IWeapon) {
    this.id = parameters.id;
    this.name = parameters.name;
    this.sideId = parameters.sideId;
    this.className = parameters.className;
    this.latitude = parameters.latitude;
    this.longitude = parameters.longitude;
    this.altitude = parameters.altitude;
    this.heading = parameters.heading;
    this.speed = parameters.speed;
    this.currentFuel = parameters.currentFuel;
    this.maxFuel = parameters.maxFuel;
    this.fuelRate = parameters.fuelRate;
    this.range = parameters.range;
    this.route = parameters.route ?? [];
    this.sideColor = convertColorNameToSideColor(parameters.sideColor);
    this.targetId = parameters.targetId;
    this.lethality = parameters.lethality;
    this.maxQuantity = parameters.maxQuantity;
    this.currentQuantity = parameters.currentQuantity;
  }
}

File: gym/blade/units/Weapon.py
import json
from typing import List, Optional
from blade.utils.colors import convert_color_name_to_side_color, SIDE_COLOR


class Weapon:
    def __init__(
        self,
        id: str,
        name: str,
        side_id: str,
        class_name: str,
        latitude: float,
        longitude: float,
        altitude: float,
        heading: float,
        speed: float,
        current_fuel: float,
        max_fuel: float,
        fuel_rate: float,  # lbs/hr
        range: float,
        route: Optional[
            List[List[float]]
        ] = None,  # Assuming route is a list of coordinates
        side_color: str | SIDE_COLOR | None = None,
        target_id: Optional[str] = None,
        lethality: float = 0.0,
        max_quantity: int = 0,
        current_quantity: int = 0,
    ):
        self.id = id
        self.name = name
        self.side_id = side_id
        self.class_name = class_name
        self.latitude = latitude
        self.longitude = longitude
        self.altitude = altitude
        self.heading = heading
        self.speed = speed
        self.current_fuel = current_fuel
        self.max_fuel = max_fuel
        self.fuel_rate = fuel_rate
        self.range = range
        self.target_id = target_id
        self.lethality = lethality
        self.max_quantity = max_quantity
        self.current_quantity = current_quantity
        self.route = route if route is not None else []
        self.side_color = convert_color_name_to_side_color(side_color)

    def get_engagement_range(self) -> float:
        return self.speed * (self.current_fuel / self.fuel_rate)

    def to_dict(self):
        return {
            "id": str(self.id),
            "name": self.name,
            "side_id": str(self.side_id),
            "class_name": self.class_name,
            "latitude": self.latitude,
            "longitude": self.longitude,
            "altitude": self.altitude,
            "heading": self.heading,
            "speed": self.speed,
            "current_fuel": self.current_fuel,
            "max_fuel": self.max_fuel,
            "fuel_rate": self.fuel_rate,
            "range": self.range,
            "target_id": str(self.target_id),
            "lethality": self.lethality,
            "max_quantity": self.max_quantity,
            "current_quantity": self.current_quantity,
            "route": self.route,
            "side_color": (
                self.side_color.value
                if isinstance(self.side_color, SIDE_COLOR)
                else self.side_color
            ),
        }
